### ISSUE TITLE
PyBind11: refactor/external API?

### ISSUE THREAD

**Issue by guacke** [GQCG/gqcp#314]
Although PyBind11 has support for Eigen (https://pybind11.readthedocs.io/en/stable/advanced/cast/eigen.html), this support seems to be limited to 'has-a/member' relationships and not 'is-a/inheritance' relationships. As such, although extending Eigen might seem a good approach to construct 'Operators', I guess that this will cause enormous problems when constructing Python bindings. 

Furthermore, code such as
```
template<typename Z = Scalar>
    static enable_if_t<std::is_same<Z, double>::value, HamiltonianParameters<double>> Random(size_t K) ...
```
is ineligible for the Python binders (and _seems_ overly complicated for what we are currently trying to achieve). Again, although C++ supports these constructs, a less elegant but 'simpler' approach should lead to easier Python bindings.

In short, this issue is centered around one question: even though C++11/14 supports certain high level abstraction concepts that we _can_ use, _should_ we use them when a simpler yet less elegant approach leads to better interoperability? The price of a possible 'back'refactor (aimed at achieving fully functioning Python bindings) should be contrasted with the possibility of an enormous gain in computational efficiency when we have fully functioning bindings.

**Comment by lelemmen** [comment]
I think it is indeed important to strive for Python interoperability, mainly because working in a Python Notebook just feels like the right workflow when working with data.

This issue might also hint at a possible `Operator` refactor, because an operator 'has a' matrix representation in an orbital basis and 'has a' matrix representation in a Fock space, not 'is a'. On the other hand, a density matrix 'is a' matrix representation, so for DMs it does make sense to inherit from `Eigen`.

As long as the resulting C++ code clearly expresses the concepts that we are implementing, I do not care about using any fancy features. If, however, a C++ paradigm (like CRTP that provides major computational speed) proves to be incompatible with any Python bindings, I feel like that part of the library shouldn't be refactored just to make Python work and we should just drop the Python bindings for that specific part of the code.

**Comment by guacke** [comment]
Agreed. If we have a solid Python bindable API (that accommodates most drivers), we can still use all C++ paradigms for 'internal' routines. As such, we should refactor the code in a such a way that the external API is relatively 'simple', but that we still have fully abstracted (C++) concepts as a backend.

With regards to a possible density matrix refactor: we could short-circuit this by building a `DensityOperator` class that 'has a' matrix representation.

**Comment by lelemmen** [comment]
Indeed. I think we both feel an important refactor is coming, that is aimed at:
- defining a Python API
- better defining our classes around an interface, which means that we should better define our objects according to what they should *behave like* (their public methods) instead of what they *are* (matrix representations)

**Comment by lelemmen** [comment]
@guacke It seems that the derived `Eigen` classes do not cause any problems for the Python bindings.
